synapse/storage/state.py

```diff
@@ -497,19 +539,24 @@ def _get_some_state_from_cache(self, group, types):
         def include(typ, state_key):
             valid_state_keys = type_to_key.get(typ, sentinel)
             if valid_state_keys is sentinel:
-                return False
+                return filtered_types is not None and typ not in filtered_types
             if valid_state_keys is None:
                 return True
             if state_key in valid_state_keys:
                 return True
             return False
 
-        got_all = is_all or not missing_types
+        got_all = is_all
+        if not got_all:
+            # the cache is incomplete. We may still have got all the results we need, if
+            # we don't have any wildcards in the match list.
+            if not missing_types and filtered_types is None:
```

Comment:
and now that we aren't returning `missing_types`, I wonder why we are bothering to build a set rather than just using a boolean. But let's just land the damn thing.